### PR TITLE
Prevent upstream error of getting property 'username' or 'password' o…

### PR DESCRIPTION
**Pull request by matthewadams**
…f undefined

The prior implementation returns `undefined`, causing line 219 to fail, because it's getting property `username` of `undefined`.  This improves the situation by at least returning an empty object.

### CODE REVIEW

**Comment by matthewadams** [comment]
@delvedor I've already signed the CLA.  I'll see about a test case when I have time.  I'm quite busy ATM.  I've added you as a collaborator in my fork at https://github.com/matthewadams/elasticsearch-js; if you would like to create a test case, create a configuration that sets `node.url` of the options object given to the `Client` constructor to a _string_ URL instead of a `URL` object.  That's enough to  reproduce the error.

**Comment by penge** [comment]
How about this?

1) Always return `{ username, password }`, not an empty object
2) Check if `node` is an object before accessing `node.url`

**Code:**

```js
function getUsernameAndPassword (node) {
  if (typeof node === 'string') {
    const { username, password } = new URL(node)
    return {
      username: decodeURIComponent(username),
      password: decodeURIComponent(password)
    }
  }

  if (typeof node === 'object' && node.url instanceof URL) { // 2
    return {
      username: decodeURIComponent(node.url.username),
      password: decodeURIComponent(node.url.password)
    }
  }

  return {
    username: undefined, // 1
    password: undefined
  }
}
```

**Output:**

```js
// string => { username: "bart", password: "123" }
console.log(
  getUsernameAndPassword('http://bart:123@localhost:9200')
);

// object => { username: "lisa", password: "456" }
console.log(
  getUsernameAndPassword({
    url: new URL('http://lisa:456@localhost:9200')  
  })
);

// something else => { username: undefined, password: undefined }
console.log(
  getUsernameAndPassword(42)
);
```

**Comment by matthewadams** [comment]
done

**Comment by penge** [comment]
@matthewadams

Can I add the test?

**Comment by matthewadams** [comment]
I just pushed what I think it ought to be -- hope you like.

File: lib/ConfigurationHelpers.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    return {
+      username: decodeURIComponent(it.username),
+      password: decodeURIComponent(it.password)
+    }
+  }
+}
+
+module.exports = {
+  getAuth
+}
```
Comment (penge) [inline lib/ConfigurationHelpers.js:37]: Let's rename this to `getCredentials()` and simplify a bit:

```js
// credentials.js

const { ConfigurationError } = require('./errors')

/**
 * Returns { username, password }, or throws ConfigurationError if node invalid
 * 
 * node can be:
 * A) 'http://abc:123@localhost:9200'
 * B) { url: new URL('http://abc:123@localhost:9200') }
 *
 * See: elasticsearch-js/test/unit/client.test.js
 */
function getCredentials (node) {
  try {
    const url = (node.url instanceof URL) ? node.url : new URL(node)
    return {
      username: decodeURIComponent(url.username),
      password: decodeURIComponent(url.password)
    }
  } catch (e) {
    throw new ConfigurationError(`invalid url in node "${node}": ${e.message}`)
  }
}

module.exports = { getCredentials }
```

Now, we can return credentials for the specific node only and move any other complexity (decision) outside, to `index.js`:

```js
// index.js

// ...
const node = Array.isArray(opts.nodes) ? opts.nodes[0] : opts.node
const credentials = getCrendentials(node)
// ...

// See also: elasticsearch-js/test/unit/client.test.js
```

**Comment by penge** [comment]
I agree. As being used at only one place, that is `index.js`, there is no really need to decouple it.

I also agree that one way to do it, would be better. If it's URL instance or string. I would probably vote for the string as it's an easier, easier to move around format. The documentation would be therefore easier too.

I think we can skip the check if the `node` is an `URL` or a `string` because that check would not be enough in case of a `string`. It can still throw a `TypeError` if the URL string is not of a valid format. We can instead "expect" the correct format to come in, and let the `URL()` constructor do all the work.

**Comment by matthewadams** [comment]
I spun the code in question out of `index.js` to make it testable.  I see no issue with spinning it out because `ConfigurationHelper.js` is down in the guts of the project.  I don't think anyone will care that it's spun out.  I'd encourage you to continue to pursue modularity like that in the name of testability.  As you design something, you should be asking yourself, "How do I test this?"  If it turns out to be hard to test, that's a yellow flag.  If it's impossible to test, that's a red flag.  Since our job as software developers is not to deliver code, but code that is _known to work_, I stand behind my work.  If you are putting something together with multiple parts, you're in a better situation if you're putting together parts that are known to work.

Also, most configuration frameworks deal with strings, not `URL` objects.  If you want to be dogmatic, yes, `node.url` should _only_ be an `instanceof URL`, but practically speaking, I don't know why you didn't make the code accept `string` or `URL` in `node`.  That would've been a good usability improvement and simplified things.  My recommendation is to deprecate `node.url` and accept `string` or `URL` in `node` (and call it `url` instead of `node`, but that's another discussion).  Remember that source code is for the humans, not the computers.

@penge & @delvedor, I have invited y'all to collaborate on my fork of this repo via the repo's settings.  I hope you appreciate my contributions of both code and advice.  Please complete the work as you see fit.

**Comment by penge** [comment]
Thank you for the lecture 😂

**Comment by matthewadams** [comment]
@delvedor I kindly invite you to complete those changes.  I don't have time right now.

**Pull request closed without merging**